_posts/2021-12-09-kafka-segment-retention.md

```diff
@@ -0,0 +1,259 @@
+By using the `DumpLogSegments` tool, it is possible to dump the `.timeindex` file content.
+
+```shell
+bin/kafka-run-class.sh kafka.tools.DumpLogSegments --deep-iteration --print-data-log --files /tmp/kafka-logs-0/__strimzi_canary-0/00000000000000000000.timeindex 
+Dumping /tmp/kafka-logs-0/__strimzi_canary-0/00000000000000000000.timeindex
+timestamp: 1638100314372 offset: 28
+timestamp: 1638100454372 offset: 56
+timestamp: 1638100594371 offset: 84
+```
+
+## Let's talk about rolling segments
+
+A new segment is rolled when some specific conditions are met.
+One condition is when the maximum segment size is reached, as specified by the configuration parameter `log.segment.bytes` (1 GiB by default).
+Another condition is based on the `log.roll.ms` or `log.roll.hours` (7 days by default) parameters. In this case, the segment is rolled when the configured time since the producer timestamp of the first record in the segment (or since the creation time if there is no timestamp) has elapsed.
+The first condition that is met will trigger the rolling of a new segment.
+It is also worth noticing that, as a not so usual use case, the producer timestamps in the records could be not ordered, so not having the older record as the first one, due to retries or the specific business logic of the producer.
+
+Another useful parameter is `log.roll.jitter.ms` which sets a maximum jitter when it's time to roll a segment,. Jitters are time intervals to avoid many segments being rolled at same time causing high contention on the disk.
+A jitter is randomly generated for each segment.
+
+The above conditions, by size or time, are the well known ones but not everyone knows that there is a third condition.
+
+A segment is rolled even when the corresponding index (or timeindex) is full.
+The index and timeindex share the same maximum size which is defined by the `log.index.size.max.bytes` configuration parameter and it is 10 MB by default.
+Let's consider the default log segment maximum size, which is 1 GiB.
+We know that because `log.index.interval.bytes` is 4096 bytes by default, an entry is added in the index every 4096 bytes of records.
+It means that for a 1 GiB segment size, 1 GiB / 4096 bytes = 262144 entries are added to the index. This equals 2 MB of index (262144 * 8 bytes).
+The default index size of 10 MB is enough to handle a segment size of 5 GiB.
+This means that reducing the index size or increasing the segment size will mean a new segment is rolled when the index is full and not when the requested segment size is reached.
+So, setting a segment size larger than 5 GiB without increasing the index size is useless, because the broker will roll a new segment when the index is full.
+
+By increasing the segment size over 5 GiB, you would also need to increase the index file size as well. Likewise, if you decide to reduce the index file size, you should probably decrease the segment size as well.
```

Comment:
Since indexes are 10MB by default, this is extreme optimization. I don't think I'd recommend lowering the index size. Indexes are trimmed once they are closed so the gains of doing so are very minimal (with a potential cost of putting a bad value).

Reply:
I agree, but I am not giving any recommendations. I am just highlighting what you should do if you decide to reduce the index file size on your own.

Reply:
To me `you should probably decrease` sounds like a recommendation. @PaulRMellor WDYT?

Reply:
@mimaison - It reads like you would probably do it to be optimal. Maybe better to remove it.

Reply:
Otherwise, maybe:  _Likewise, if you decide to reduce the index file size, it is possible that you might want to decrease the segment size proportionately._

Reply:
@mimaison I would leave the case but using the sentence proposed by Paul instead of removing it. Wdyt?